### PR TITLE
ValueFactory (nonce)

### PR DESCRIPTION
**Pull request by rafalkrupinski**
These commits introduce ValueFactory interface and a simple implementation for it, based on System.currentTimeMillis().

The purpose is to calculate nonce just before sending it, both in a single synchronized block.

There are simple unit tests, I have also tested it on a live exchange. https://github.com/rafalkrupinski/XChange/tree/nonce-kraken

also see https://github.com/timmolter/XChange/issues/525

### CODE REVIEW

**Comment by mmazi** [comment]
How urgent is this? I'm currently on vacation, but had a brief look at the PR and it seems there are two issues with it that I'd like to resolve first before merging:
1. It doesn't seem intuitive to me that, in general, having a ValueFactory parameter implies synchronisation. This is ok in the particular nonce case, but I don't see it in general.
2. Perhaps ValueFactroy can somehow be merged with ParamsDigest, or at least some parts of the code that calls the ValueFactory method.

I think this pull request solves the nonce issue, but should probably be generalized a bit differently to find some other uses too.

**Comment by timmolter** [comment]
Well, unfortunately I pulled a PR into XChange based on the assumption that you already merged the PR and it was in a jar on Sonatype, and now our Kraken module is broken. I guess I could revert it, but I don't know how. Is reverting hard to do with git? @rafalkrupinski can comment on your other comments.

**Comment by mmazi** [comment]
Sorry, my bad.. You can do a git reset and push into github, but that's changing history which will cause trouble to anyone who has pulled during this time (eg. Travis, or github itself, which is where I messed up hoping that no one has pulled during the minute my mistaken merge was on github, but of course github caught it). So I wouldn't recommend this. I suggest you do a git revert of the merge commit.

**Comment by rafalkrupinski** [comment]
Ad 1 - I couldn't think of any other use case than to enforce generation in a block synchronized with sending.
Ad 2 - I don't think they can be merged, MessageDigest must be called last to pick up any generated values.

The matter is, is the public API OK? If so, maybe Tim doesn't have to revert.

**Comment by mmazi** [comment]
I might want to change the public API. Also, I'd like to have some test that demonstrate that this actually solves the potential race condition mentioned in https://github.com/timmolter/XChange/issues/525 (eg. by running several threads etc.)

If I understand https://github.com/timmolter/XChange/issues/525 correctly, this is not urgent since it's meant to solve a _potential_ problem? In this case I'd really prefer to take my time and look into this issue when I have the time, in two weeks or so.

**Comment by mmazi** [comment]
@timmolter Ugh, [reverting a merge](http://git-scm.com/blog/2010/03/02/undoing-merges.html) is not so easy... Or, it is easy, but will leave some loose ends, namely merging the PR again will be confusing.

So I'll just accept the ValueFactory PR as it is now so that XChange will work. I'll refactor it (in rescu and in XChange at the same time) later on if I still feel like it.

Thanks for the contribution!

**Comment by timmolter** [comment]
@mmazi 
Yeah a revert is tricky. I was going to revert it manually and push the changes back. But if you will merge the PR now, that'd be better anyway.

I understand that you'd like unit tests to demonstrate the issue, and that you want to take your time. Let's look at it together when you get back from vacation. 

It will be hard to design a test case. From what I've heard from others, nonce mixups happen sometimes. If the nonce calculation is delayed to the last moment before the message is sent in rescu, it sounds reasonable that there would be less nonce issues. I'll admit though that I haven't thought deeply about it.

Aside form that though, the ValueFactory is a neat way to abstract nonce generation.

File: src/main/java/si/mazi/rescu/RestInvocationHandler.java [hunk 1]
```diff
@@ -89,27 +90,47 @@ public RestInvocationHandler(Class<?> restInterface, String url, ClientConfig co
...
+        HttpURLConnection connection;
+        if (hasValueGenerator(args)) {
+            synchronized (this) {
+                connection = prepareAndInvoke(args, methodMetadata);
+            }
+        } else {
+            connection = prepareAndInvoke(args, methodMetadata);
+        }
+
+        return receiveAndMap(methodMetadata, connection);
```
Comment (sutra) [inline src/main/java/si/mazi/rescu/RestInvocationHandler.java:103]: Only the sending is in the synchronized block, but the receiving is out of the block, can this ensure the requests reach the server side orderly? For instance, if we have 2 requests, any possible that the second request reaches the server side before the first request?

File: src/main/java/si/mazi/rescu/RestInvocationHandler.java [hunk 2]
```diff
@@ -89,27 +90,47 @@ public RestInvocationHandler(Class<?> restInterface, String url, ClientConfig co
 
     public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
         RestMethodMetadata methodMetadata = getMetadata(method);
-        
+
+        HttpURLConnection connection;
+        if (hasValueGenerator(args)) {
+            synchronized (this) {
```
Comment (sutra) [inline src/main/java/si/mazi/rescu/RestInvocationHandler.java:96]: Use the ValueFactory instance as the synchronized lock?

Comment (rafalkrupinski) [inline src/main/java/si/mazi/rescu/RestInvocationHandler.java:96]: This is probably a good idea.

File: src/main/java/si/mazi/rescu/RestInvocationHandler.java (hunk 1 above)
Comment (rafalkrupinski) [inline src/main/java/si/mazi/rescu/RestInvocationHandler.java:103]: We're trying to find a balance between the highest possible communication speed and ensuring the messages are received in the same order.
Ideally we would send a message as soon as the previous one was received by the server, but before it's processed, unfortunately we don't know when exactly is that.

Reading data in the synchronized block would make the communication with the exchange fully serialized (single threaded). I'm not sure if that's what we want. Is it?

We could add flush() in the block to make it more probable that it is sent in the the correct order.

**Pull request merged**